### ISSUE TITLE
Allow injections in overwritten methods 

### ISSUE THREAD

**Issue by Yeregorix** [SpongePowered/Mixin#251]
I need to `@Inject` some code at the end of a specific mc method which is already `@Overwritten` by another mod (which uses Mixin) but it's currently not possible. It makes sense when the injection point is at specific opcode because the whole method was changed and the injection point may have moved or totally disappeared. However I can't see why we can't inject at the HEAD or the RETURN point(s) of the method. This would be really useful when the injection is just a kind of listener which don't care if the target method was modified or not.

**Comment by Mumfrey** [comment]
Indeed, the reason for prohibiting this is that injections are usually tied to the method structure. However `HEAD` and `RETURN` are not actually dependent on this behaviour so it would make sense to allow an exception to this. Good suggestion.

**Comment by Barteks2x** [comment]
Could there be anyway to specifically allow injecting into overwritten methods when when it's not HEAD or RETURN? This could be useful for solving compatibility issues between different mods that use mixin, where one overwrites a method, and another one injects into it. Or is there already a way to make that work? Or a completely different way to solve the issue?

**Comment by Mumfrey** [comment]
I think the only way it could realistically be allowed is if some extra metadata were added to the injector to say specifically _"this injector is intended to inject into a **specific** overwritten method from `MixinBlah`"_ because otherwise there's no way to realistically be sure that the injector is going to behave as intended.

The thought process going like this:

* In general, inject into a target method will succeed based on its criteria, simple
* If a method is overwritten, the structure might have significantly changed, so injectors are inhibited so that they don't do unpredictable things
* If an injector is `HEAD` or `RETURN`, we can make an exception because the structure of the method doesn't affect the contract of those injectors, thus they should behave the same
* _If you need to inject into a **specific, anticipated** overwrite, this should only be possible if you **explicitly specify** the fully qualified name of the mixin owning the overwrite, indicating that you know you're injecting into an overwritten method. Or some other method of indicating this awareness._

I was planning to add an accessor in the base class `InjectionPoint` something like `canInjectIntoOverwrite()` so that the information on injection points which can and cannot inject into overwrites is stored with the injection point. This would also make it possible to acheive the latter goal using custom InjectionPoint subclass, which might be a better method since then it puts more control back into the hands of the mixin author, at the expense of some minor additional complexity. This might be a good trade-off.

**Comment by Mumfrey** [comment]
So looking into this, it looks like I already took this situation into account with the priority system. If a mixin has higher priority than the one which merged the method then it should be able to inject into the method just fine.

**Comment by Yeregorix** [comment]
If the mixin containing the `@Inject` has an higher priority than the mixin containing the `@Overwrite` then the code injected by the first one is erased by the second one. If the mixin containing the `@Inject` has an lower priority then it can't be injected at all. Currently, from what I know, there is no way to `@Inject` after an `@Overwrite`, as explained above. Making an exception for `HEAD` or `RETURN` injection points or allowing to specify the fully qualified name of the mixin owning the `@Overwrite` seem to be good solutions for me. ^^

**Comment by Mumfrey** [comment]
> If the mixin containing the @Inject has an higher priority than the mixin containing the @Overwrite then the code injected by the first one is erased by the second one.

How? If the injector is in a higher priority mixin then why would it overwrite the previous overwrite?

> If the mixin containing the @Inject has an lower priority then it can't be injected at all.

Correct, this is intentional.

> Currently, from what I know, there is no way to @Inject after an @Overwrite, as explained above.

Yes there is, the injector mixin just needs a higher priority than the overwriting mixin.

> Making an exception for HEAD or RETURN injection points or allowing to specify the fully qualified name of the mixin owning the @Overwrite seem to be good solutions for me.

This is true, for an exception to the priority rule, but your implication is that there is currently *no* way to inject into an overwritten method, but this isn't true.

**All** methods are merged before injectors are even inspected, so if two mixins target the same class then both of their methods will be added. In the case of an overwrite the higher priority "wins" (unless the lower priority one is `@Final`) but for everything else, including injectors (which are conformed so that injectors with the same name will not overlap) will be merged. The injectors are processed at a late stage in application so that the class's structure is "finalised" before the injectors run.

I'm not really sure what you're getting at here. It seems like this functionality is already covered adequately via priority with the exception of the enhancement mentioned.

**Comment by Yeregorix** [comment]
I thought after some tests that @Overwrite was erasing all injections but I just did better tests and I've found that I was wrong. Sorry for that ^^'. Anyway the enhancement mentioned is still a good idea. 👍
